docs/supported_formats.rst

```diff
@@ -0,0 +1,586 @@
+Supported Formats
+=================
+
+ImageIO reads and writes images by deligating your request to one of many
+backends, e.g., pillow, ffmpeg, tifffile, etc. Each backend supports
+its own set of formats, which is how ImageIO manages to support so many
+of them.
+
+Below you can find a (non-exhaustive) list of formats that ImageIO supports
+together with the backend supporting it. This list is non-exhaustive, because it
+only lists the formats we were aware of at the time the docs were written. If
+any backend introduces support for new formats they, too, will be supported but
+our docs may take time to update and reflect this. If in doubt, check the docs
+of the respective backend.
+
+Format List
+-----------
+Each backend listed here is listed using the format string you would use to call
```

Comment:
Is this file still auto-generated? I see code-gen code being removed, but no new code to replace it. I hope it's a manual document, because I suspect that will be painful to maintain.

Reply:
You hope it's a manual document or you hope its _not_ a manual document? But to answer your question either way: At the moment it is a manual document.

Moving forward, I think it will become a little easier, since a lot of the formats are `XXX-FI` or `XXX-PIL`, and in the new API all of these will become `plugin="freeimage"` or `plugin="pillow"`, at which point we could think of removing the Format List (by extension) and just keep a format list by backend and ask people to Ctrl+F the extension they are looking for. Then again it might be nice to have a long list of extensions/formats, because it looks impressive.

Reply:
Oh, and some of the codegen is no covered by autosummary/autodoc. I also removed all the template strings from the extension and placed them into dedicated template files.

Reply:
Yeah I meant *not* a manual document, sorry.

> [...] in the new API all of these will become `plugin="freeimage"` or `plugin="pillow"`.

mmm what would that look like? I imagine that many formats still have their own specific keyword argument (e.g. compression for png, exif rotation for jpg, etc, etc.).

Reply:
In either case, this list seems like it's easy to generate from the information that is already available. It being manual will mean having to keep things in sync, and discrepancies will creep in at some point.

Reply:
> mmm what would that look like? I imagine that many formats still have their own specific keyword argument

@almarklein Sure, you can just append those to the call and they will get sorted out by the plugin or the backend, e.g., `iio.imread("some.jpg", plugin="pillow", rotate=False)`. If there are a lot of backend-specific ones like there are for ffmpeg a plugin can specify `**kwargs` in its implementation of the interface and then forward them appropriately or throw an Exception for unsupported ones.

If you were/are wondering how it will look like on the documentation side then that will depend on the individual plugin/backend. Each plugin has its module parsed by autosummary/autodoc and then it can decide how it wants to be structured. At the moment all plugins are single page and follow numpydoc style where possible, but it doesn't have to be like this. You could have a `Format Specific Parameters` section where you list the parameters that are unique to a format if there are many, or you can simply list them in a single `Parameters` list and flag them appropriately (what I've been doing so far since it's only a few parameters if any). 

The main point is that if you use a plugin, say `plugin=dicom`, then the DICOM plugin API reference is _the_ place to search for any parameters; format-specific or otherwise. No more need to start digging through Supported Formats (this info doesn't really belong there anyway). Supported Formats will instead list what format is read by which plugin; if you wonder "(Can imageio read)/(Does imageio support) X " (for X = RTSP streams :D https://github.com/imageio/imageio/issues/658) then you go to Supported Formats, Crtl + F for the extension or format name, and check which plugin you need to install.

If you _need_ to read a specific file with a specific plugin using as a specific format and the format is - for whatever reason - different from the format that the backend detects (via extension magic number or otherwise), then you need to check the plugin docs regarding the keyword that forces this particular backend to interpret the specific file as a specific format. This does seem rather niche, so I think it is okay that the call is a bit more sophisticated (`iio.imread(specific_file, plugin="specific_plugin", kwarg_enforcing_format=...)`.

---

> this list seems like it's easy to generate from the information that is already available

Yes, partially. In the legacy plugin manager a plugin must explicitly register formats, which I guess we could tap into to populate the list. The new plugin manager currently doesn't care about formats (by design) and sees them as more of an optional optimization (https://github.com/imageio/imageio/issues/641) than mandatory information. We will need to write some kind of dict/json file/thing for https://github.com/imageio/imageio/issues/641 that signals which plugins to prioritize based on file extension. We could again use this to populate the list.

What we can't easily automate is the association of an extension with its full format name (and link to additional documentation).  I think this information is important because it clarifies what an extension is and it (may) de-alias formats that share the same extensions. The only way I can think of maintaining this is to template the list and then simply populate pre-existing entries based on the extension-plugin pairs we get from somewhere. We will still have to manually add new formats and we won't gain guarantees on completeness (since backends might introduce new formats without us being aware for a while). The library will support those new formats immediately without changes, but the docs will lag behind. Considering this, automation didn't offer enough bang for the buck imo, so I just wrote a manual one.

That said, I'm happy to automate if you have a better idea.

Reply:
> If you were/are wondering how it will look like on the documentation side then that will depend on the individual plugin/backend. 

Yes, I was wondering mostly about the docs. Ok, I think I see what you mean. So the Pillow plugin will be a relatively large page describing the parameters for generic formats, and a few sections for png, jpg, etc.? I hadn't thought about that, but it sounds good. Would be nice, though, if the links in the list of supported formats would bring the user to the relevant section (Eg for png, it lists at least pillow, freeimage, and if you click on pillow you get to the section on the pillow page describing the png format params). I guesst this could be as simple as adding the format at the end of the link ("/plugins/pillow.html#png"), and making sure that the doc page has `<a name='png' />`.

----

> What we can't easily automate is the association of an extension with its full format name (and link to additional documentation). 

So perhaps we should add some information to the plugin classes so that we can? ;)  Maintaining some meta-data at the plugins is easier than maintaining a large manual list.

Reply:
> Would be nice, though, if the links in the list of supported formats would bring the user to the relevant section

Nice idea! I'm sold :D This might be a tad iffy to maintain if we automatically generate the list - only a few plugins will have the need for this -, but we can see.

> So perhaps we should add some information to the plugin classes so that we can? ;) 

What kind of info did you have in mind and where should it live? In my mind that would be a description of each format supported by a plugin/backend. This would mean a lot of replication if multiple formats can read the same format (tiff, jpg, png, ...) that needs to be kept in sync. I guess I am misunderstanding what you have in mind.

Reply:
Haha, ok, I think I see what you mean. You have a format, say PNG, and with it a short description and a link to the wiki page about that format, and then a link to each plugin that supports that format. So maybe we should create a definition of formats, if only for the sake of documentation. Something like:

```
format_meta_data = {
     "png": {
        "description": """ PNG is a lossless image format, blabla
            ....
        """,
       "plugins": ["pillow", "freeimage"], 
}
```
or
```
class PngFormat:
   """ PNG is a lossless bla bla
   ..
   """
   plugins = "pillow", "freeimage" 
```

And plugins that specify what formats they support based on the key ("png" in this case).

Maybe we could also specify common extension here, instead of on the plugins. mmmm

Reply:
> So maybe we should create a definition of formats, if only for the sake of documentation. [...] And plugins that specify what formats they support based on the key ("png" in this case).

Just to poke the bear: If we only do it for the docs, why not use a `_templates/format.rst` file instead of a meta dict and have it populated by our extension or autodoc? This would give us much easier control over the formatting and allow IDEs with rst support to perform syntax and cross-ref checks. We can also place the description into it directly since it won't be used by any other part of our codebase.

> Maybe we could also specify common extension here, instead of on the plugins.

How do you mean? I thought the (dict keys)/(class names) would be based on extensions? Or are you thinking of basing them on a format's full name and then add extensions as a metadata key?

Reply:
> why not use a _templates/format.rst file

I have never worked with rst templates, so yeah, if that works, great!

> How do you mean? I thought the (dict keys)/(class names) would be based on extensions? Or are you thinking of basing them on a format's full name and then add extensions as a metadata key?

So right now (in the old system) a plugin class has an `extensions` attribute that is used to pre-filter possible plugins. We mean to bring something like this back. But instead of specifying extensions, we could specify format names, and declare what extensions are common to this format in a different (central) place. If we'd go with the rst solution, thus would make much less sense though, I guess :)